src/tools/perf/perftest_mad.c

```diff
@@ -0,0 +1,725 @@
+        } else if (errno == EINVAL) {
+            /* EINVAL: no length info */
+            ref_len *= 2; /* 'len' has become invalid */
+        } else {
+            /* ENOSPC: needed length is set */
+            ucs_assert(errno == ENOSPC);
+            ref_len = len;
+        }
+    }
+
+    if (perftest_mad_get_remote_port(umad, remote_port) != UCS_OK) {
+        status = UCS_ERR_IO_ERROR;
+        goto terminate;
+    }
+
+    len -= IB_VENDOR_RANGE2_DATA_OFFS;
+    if (len <= 0) {
+        if (user_mad->status == ETIMEDOUT) {
+            ucs_error("MAD: receive: Remote unreachable");
+            status = UCS_ERR_UNREACHABLE;
+        } else {
+            status = UCS_ERR_OUT_OF_RANGE;
+        }
+        goto terminate;
+    }
+
+    ret = umad_status(umad);
+    if (ret) {
+        ucs_warn("MAD: receive: Status failure: %d", ret);
+        status = UCS_ERR_REJECTED;
+        goto terminate;
+    }
+
+    status = UCS_OK;
+
+    data = umad_get_mad(umad);
+    data = UCS_PTR_BYTE_OFFSET(data, IB_VENDOR_RANGE2_DATA_OFFS);
+    if (perftest_mad_is_ping(data, len)) {
+        ucs_debug("PONG");
+        goto ping_retry;
+    }
+
+    if (len > *avail) {
+        status = UCS_ERR_MESSAGE_TRUNCATED;
+        len    = *avail;
+    }
+    memcpy(buffer, data, len);
+    *avail = len;
+
+terminate:
+    free(umad);
+    return status;
+}
+
+static ucs_status_t
+perftest_mad_recv_from_remote(perftest_mad_rte_group_t *rte_group, void *buffer,
+                              size_t *avail, const ib_portid_t *target_port)
+{
+    ucs_status_t status     = UCS_ERR_IO_ERROR;
+    ib_portid_t remote_port = {
+        .lid = 0
+    };
+    size_t size;
+
+    while ((remote_port.lid == 0) || (remote_port.lid != target_port->lid)) {
+        size            = *avail;
+        remote_port.lid = 0;
+
+        status = perftest_mad_recv(rte_group, buffer, &size, &remote_port);
+        if (status == UCS_ERR_NO_PROGRESS) {
+            ucs_debug("PING");
+            status = perftest_mad_ping(rte_group);
+            ucs_assertv(status == UCS_OK, "status: %s",
+                        ucs_status_string(status));
+            remote_port.lid = 0;
+            continue;
+        }
+        if (status == UCS_ERR_UNREACHABLE) {
+            return status;
+        }
+    }
+    ucs_debug("MAD: recv packet size:%zu/%zu", size, *avail);
+    *avail = size;
+    return status;
+}
+
+static unsigned rte_mad_group_size(void *rte_group)
+{
+    return 2;
+}
+
+static unsigned rte_mad_group_index(void *rte_group)
+{
+    return !((perftest_mad_rte_group_t*)rte_group)->is_server;
+}
+
+static ucs_status_t
+perftest_mad_recv_magic(perftest_mad_rte_group_t *group, unsigned value)
+{
+    unsigned snc = 0;
+    size_t size  = sizeof(snc);
+    ucs_status_t status;
+
+    status = perftest_mad_recv_from_remote(group, &snc, &size,
+                                           &group->dst_port);
+    ucs_debug("recv magic: snc 0x%08x, value 0x%08x", snc, value);
```

Comment:
move this debug message after checking status is ok (if status not ok, no need to print snc)
also, what is the meaning of the name "snc"

Reply:
i assume this is for "sync" since the same naming was used for the existing sock rte, i renamed it to index, moved the log after status checking